vmm/src/device_manager.rs

```diff
@@ -1414,93 +1417,97 @@ impl DeviceManager {
-                            fs_sock,
-                            &fs_cfg.tag,
-                            fs_cfg.num_queues,
-                            fs_cfg.queue_size,
-                            cache,
-                        )
-                        .map_err(DeviceManagerError::CreateVirtioFs)?,
-                    ));
+                let mmap_region = MmapRegion::build(
+                    None,
+                    fs_cache as usize,
+                    libc::PROT_NONE,
+                    libc::MAP_ANONYMOUS | libc::MAP_PRIVATE,
+                )
+                .map_err(DeviceManagerError::NewMmapRegion)?;
+                let host_addr: u64 = mmap_region.as_ptr() as u64;
 
-                    devices.push((
-                        Arc::clone(&virtio_fs_device) as VirtioDeviceArc,
+                let mem_slot = self
+                    .memory_manager
+                    .lock()
+                    .unwrap()
+                    .create_userspace_mapping(
+                        fs_guest_addr.raw_value(),
+                        fs_cache,
+                        host_addr,
                         false,
-                        None,
-                    ));
+                        false,
+                    )
+                    .map_err(DeviceManagerError::MemoryManager)?;
+
+                let mut region_list = Vec::new();
+                region_list.push(VirtioSharedMemory {
+                    offset: 0,
+                    len: fs_cache,
+                });
+
+                Some((
+                    VirtioSharedMemoryList {
+                        host_addr,
+                        mem_slot,
+                        addr: fs_guest_addr,
+                        len: fs_cache as GuestUsize,
+                        region_list,
+                    },
+                    mmap_region,
+                ))
+            } else {
+                None
+            };
 
-                    let migratable = Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>;
-                    let id = migratable.lock().unwrap().id();
-                    self.migratable_devices.push((id, migratable));
-                }
+            let virtio_fs_device = Arc::new(Mutex::new(
+                vm_virtio::vhost_user::Fs::new(
+                    fs_sock,
+                    &fs_cfg.tag,
+                    fs_cfg.num_queues,
+                    fs_cfg.queue_size,
+                    cache,
+                )
+                .map_err(DeviceManagerError::CreateVirtioFs)?,
+            ));
+
+            self.add_migratable_device(Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>);
+
+            Ok((
+                Arc::clone(&virtio_fs_device) as VirtioDeviceArc,
+                false,
+                None,
+            ))
+        } else {
+            Err(DeviceManagerError::NoVirtioFsSock)
+        }
+    }
+
+    fn make_virtio_fs_devices(
```

Comment:
Have you tested creating multiple fs devices?

Reply:
Actually no, I can have a try, but right now I don't know why the CI is not passing...

Reply:
I'm seeing the same issues in our project, the second virtio-fs device won't mount and the guest kernel reports that the tag was not found.

Reply:
Ok the PR is fixed! I added one more patch (a oneliner changing quite a lot of things) to make everything work. I'm glad I factorized the integration test because it made me realize we had a problem as virtio-fs was not working after some memory was hotplugged.

About multiple devices, I've just tested that and it works fine for me. You need to run multiple backends and then hotplug them one after the other. If you mount them in different locations on the host it should work for you too :)

Reply:
Seems to be working but I'm hitting a connection limit wall even after bumping the connection limit constant up... I'm trying to add 10 or so FSes, and the first 8 or so seem to work

Reply:
What do you mean by connection limit?

I'm curious why you need so many virtio-fs devices to be passed into a single VM. Are you passing one virtio-fs device per container that you run into the VM?
I'm asking because you can bypass this limitation by having one single virtio-fs device shared across all containers in the same VM.

One more thing, I never tried so many `virtio-fs` devices, but there's a hard limit set at 32 PCI devices since we emulate only one PCI bus with Cloud-Hypervisor.

Reply:
That makes sense, we'll just stick to one virtio-fs then per VM.